File: frontend/js/ops.js
(function (global) {
  "use strict";

  const REWRITE_FIELDS = new Set([
    "title", "subtitle", "kicker", "body", "text", "intro", "desc", "caption", "from", "no", "bridge", "role", "progress", "contact", "label",
  ]);

  const PAGE_TYPES = new Set([
    "cover", "chapter", "points", "timeline", "card", "quote", "compare", "summary", "ending", "composite", "photo", "gallery", "free",
  ]);

  function clone(v) {
    return JSON.parse(JSON.stringify(v));
  }

  function listKeyForPage(page) {
    if (!page || typeof page !== "object") return null;
    const t = page.type || "";
    if (t === "points" || t === "summary") return "items";
    if (t === "timeline") return "steps";
    if (t === "compare") return "rows";
    if (t === "free") {
      if (Array.isArray(page.bullets) && page.bullets.length) return "bullets";
      if (Array.isArray(page.paragraphs) && page.paragraphs.length) return "paragraphs";
      return "bullets";
    }
    if (t === "card") return "tips";
    if (t === "ending") return "cta";
    return null;
  }

  function markUserEdited(page) {
    if (!page || typeof page !== "object") return page;
    return { ...page, _user_edited: true };
  }

  function normalizeOp(raw) {
    if (!raw || typeof raw !== "object") return null;
    const name = String(raw.op || "").trim();
    const allowed = new Set([
      "setDensity", "setFontScale", "rewriteField", "deleteItem", "updateItem", "setPageType",
      "setPointsLayout", "setPageLayoutVariant", "setLayoutRecipe",
      "moveItem", "insertItem", "setImage", "setImageFocus", "patchPath",
    ]);
    if (!allowed.has(name)) return null;
    const out = { op: name };
    if (name === "setDensity") {
      const d = String(raw.density || "").toLowerCase();
      if (!["compact", "normal", "airy"].includes(d)) return null;
      out.density = d;
    } else if (name === "setFontScale") {
      const fs = Number(raw.fontScale);
      if (!Number.isFinite(fs)) return null;
      out.fontScale = Math.max(0.65, Math.min(1.4, fs));
    } else if (name === "rewriteField") {
      const field = String(raw.field || "").trim();
      if (!REWRITE_FIELDS.has(field)) return null;
      out.field = field;
      out.value = raw.value == null ? "" : String(raw.value);
    } else if (name === "deleteItem" || name === "updateItem" || name === "moveItem" || name === "insertItem") {
      const idx = parseInt(raw.index, 10);
      const from = parseInt(raw.from, 10);
      const to = parseInt(raw.to, 10);
      if (name === "moveItem") {
        if (!(from >= 1) || !(to >= 1)) return null;
        out.from = from;
        out.to = to;
      } else if (name === "insertItem") {
        if (!(idx >= 1)) return null;
        out.index = idx;
        out.item = raw.item;
      } else {
        if (!(idx >= 1)) return null;
        out.index = idx;
      }
      const lst = String(raw.list || "").trim();
      if (lst) out.list = lst;
      const lp = String(raw.listPath || "").trim();
      if (lp) out.listPath = lp;
      if (name === "updateItem") {
        ["head", "body", "value", "text", "time", "label"].forEach((k) => {
          if (raw[k] != null) out[k] = String(raw[k]);
        });
        const vi = parseInt(raw.valueIndex, 10);
        if (vi >= 1) out.valueIndex = vi;
      }
    } else if (name === "setImage") {
      out.slot = String(raw.slot || "image");
      out.url = String(raw.url || "");
      if (raw.focus) out.focus = raw.focus;
    } else if (name === "setImageFocus") {
      out.slot = String(raw.slot || "image");
      out.x = Math.max(0, Math.min(100, Number(raw.x) || 50));
      out.y = Math.max(0, Math.min(100, Number(raw.y) || 50));
    } else if (name === "setPageType") {
      const t = String(raw.type || "");
      if (!PAGE_TYPES.has(t)) return null;
      out.type = t;
    } else if (name === "setPointsLayout") {
      const layout = String(raw.layout || "").toLowerCase();
      if (!["ledger", "cards"].includes(layout)) return null;
      out.layout = layout;
    } else if (name === "setPageLayoutVariant") {
      const variant = String(raw.variant || raw.layoutVariant || "").trim();
      if (!variant) return null;
      out.variant = variant;
    } else if (name === "setLayoutRecipe") {
      const recipeId = String(raw.recipeId || "").trim();
      if (!recipeId) return null;
      out.recipeId = recipeId;
    } else if (name === "patchPath") {
      const p = String(raw.path || "").trim();
      if (!p) return null;
      out.path = p;
      out.value = raw.value == null ? "" : raw.value;
    }
    return out;
  }

  function applyDensity(page, density) {
    const out = { ...page, density };
    const fs = density === "compact" ? 0.88 : density === "airy" ? 1.12 : 1;
    out.fontScale = fs;
    out.style = { ...(page.style || {}), density, fontScale: fs };
    return out;
  }

  function listPathKey(op) {
    if (op.listPath) return { path: op.listPath, nested: true };
    if (op.list && String(op.list).includes("[")) return { path: op.list, nested: true };
    if (op.list) return { path: op.list, nested: false };
    return null;
  }

  function readList(page, spec) {
    if (!spec) return null;
    if (spec.nested) {
      const arr = global.XHSFieldPath.getAtPath(page, spec.path);
      return Array.isArray(arr) ? arr : null;
    }
    return page[spec.path];
  }

  function writeList(page, spec, items) {
    if (!spec) return page;
    if (spec.nested) {
      global.XHSFieldPath.setAtPath(page, spec.path, items);
      return page;
    }
    page[spec.path] = items;
    return page;
  }

  function applyPageOps(page, ops) {
    let out = clone(page || { type: "card", title: "内容", body: "" });
    const applied = [];
    for (const raw of ops || []) {
      const op = normalizeOp(raw);
      if (!op) continue;
      const name = op.op;
      if (name === "setDensity") {
        out = applyDensity(out, op.density);
      } else if (name === "setFontScale") {
        out.fontScale = op.fontScale;
        let dens = out.density || "normal";
        if (op.fontScale <= 0.9) dens = "compact";
        else if (op.fontScale >= 1.1) dens = "airy";
        out.density = dens;
        out.style = { ...(out.style || {}), density: dens, fontScale: op.fontScale };
      } else if (name === "rewriteField") {
        out[op.field] = op.value;
      } else if (name === "patchPath") {
        global.XHSFieldPath.setAtPath(out, op.path, op.value);
      } else if (name === "deleteItem") {
        const spec = listPathKey(op) || (listKeyForPage(out) ? { path: listKeyForPage(out), nested: false } : null);
        if (!spec) continue;
        const items = [...(readList(out, spec) || [])];
        const idx = op.index;
        if (idx < 1 || idx > items.length) continue;
        items.splice(idx - 1, 1);
        writeList(out, spec, items);
      } else if (name === "insertItem") {
        const spec = listPathKey(op) || (listKeyForPage(out) ? { path: listKeyForPage(out), nested: false } : null);
        if (!spec) continue;
        const key = spec.nested ? spec.path.split(".").pop() : spec.path;
        const items = [...(readList(out, spec) || [])];
        const idx = Math.min(op.index, items.length + 1);
        const item = op.item || (key === "items" || key === "steps" ? { head: "新条目", body: "" } : "新条目");
        items.splice(idx - 1, 0, item);
        writeList(out, spec, items);
      } else if (name === "moveItem") {
        const spec = listPathKey(op) || (listKeyForPage(out) ? { path: listKeyForPage(out), nested: false } : null);
        if (!spec) continue;
        const items = [...(readList(out, spec) || [])];
        const from = op.from - 1;
        const to = op.to - 1;
        if (from < 0 || from >= items.length || to < 0 || to >= items.length) continue;
        const [moved] = items.splice(from, 1);
        items.splice(to, 0, moved);
        writeList(out, spec, items);
      } else if (name === "updateItem") {
        const spec = listPathKey(op) || (listKeyForPage(out) ? { path: listKeyForPage(out), nested: false } : null);
        if (!spec) continue;
        const items = [...(readList(out, spec) || [])];
        const idx = op.index;
        if (idx < 1 || idx > items.length) continue;
        const cur = items[idx - 1];
        const newText = op.body || op.value || op.text;
        if (typeof cur === "object" && cur) {
          const nxt = { ...cur };
          ["head", "body", "value", "text", "time", "label"].forEach((k) => {
            if (op[k] != null) nxt[k] = op[k];
          });
          if (op.valueIndex != null) {
            const vi = parseInt(op.valueIndex, 10) - 1;
            if (vi >= 0) {
              const vals = Array.isArray(nxt.values) ? [...nxt.values] : [];
              while (vals.length <= vi) vals.push("");
              vals[vi] = op.value != null ? op.value : newText != null ? newText : vals[vi];
              nxt.values = vals;
            }
          }
          items[idx - 1] = nxt;
        } else if (newText != null) {
          items[idx - 1] = newText;
        }
        writeList(out, spec, items);
      } else if (name === "setImage") {
        const slot = op.slot || "image";
        if (slot.startsWith("images[")) {
          const m = slot.match(/images\[(\d+)\]/);
          if (m) {
            const i = parseInt(m[1], 10);
            const imgs = [...(out.images || [])];
            while (imgs.length <= i) imgs.push("");
            imgs[i] = op.url;
            out.images = imgs;
          }
        } else {
          out.image = op.url;
        }
        if (op.focus) {
          out.imageFocus = op.focus;
        }
      } else if (name === "setImageFocus") {
        out.imageFocus = { x: op.x, y: op.y };
      } else if (name === "setPageType") {
        out.type = op.type;
      } else if (name === "setPointsLayout") {
        if ((out.type || "points") !== "points") continue;
        out.pointsStyle = op.layout;
        out.layoutVariant = op.layout;
        const styleObj = out.style && typeof out.style === "object" ? { ...out.style } : {};
        styleObj.pointsStyle = op.layout;
        styleObj.layoutVariant = op.layout;
        out.style = styleObj;
      } else if (name === "setPageLayoutVariant") {
        out.layoutVariant = op.variant;
        const styleObj = out.style && typeof out.style === "object" ? { ...out.style } : {};
        styleObj.layoutVariant = op.variant;
        out.style = styleObj;
        if (out.type === "points" && (op.variant === "ledger" || op.variant === "cards")) {
          out.pointsStyle = op.variant;
          styleObj.pointsStyle = op.variant;
        }
      } else if (name === "setLayoutRecipe") {
        out.layoutRecipe = op.recipeId;
        const defaults = (global.XHSLayoutCatalog && global.XHSLayoutCatalog.defaultsFor(op.recipeId)) || {};
        if (defaults.density) out = applyDensity(out, defaults.density);
        if (defaults.pointsStyle && out.type === "points") {
          out.pointsStyle = defaults.pointsStyle;
          const styleObj = out.style && typeof out.style === "object" ? { ...out.style } : {};
          styleObj.pointsStyle = defaults.pointsStyle;
          out.style = styleObj;
        }
      }
      applied.push(op);
    }
    return { page: markUserEdited(out), applied, summary: applied.map((o) => o.op).join(" · ") };
  }

  function blankPage(type) {
    const t = type || "card";
    const map = {
      cover: { type: "cover", title: "新封面", subtitle: "副标题" },
      ending: { type: "ending", title: "结尾", desc: "行动号召", cta: ["收藏备用"] },
      points: { type: "points", title: "要点", items: [{ head: "1", body: "内容" }] },
      timeline: { type: "timeline", title: "步骤", steps: [{ time: "Day1", head: "步骤", body: "说明" }] },
      card: { type: "card", title: "卡片", body: "正文" },
      quote: { type: "quote", text: "金句" },
      compare: { type: "compare", title: "对比", cols: [{ head: "维度" }, { head: "A", tone: "neg" }, { head: "B", tone: "pos" }], rows: [{ label: "项", values: ["", ""] }] },
      summary: { type: "summary", title: "总结", items: ["要点一"] },
      free: { type: "free", title: "展开", paragraphs: ["正文"] },
      photo: { type: "photo", title: "配图", body: "说明" },
      gallery: { type: "gallery", title: "图集", images: [] },
    };
    return map[t] || map.card;
  }

  function applyDocumentOps(data, ops) {
    const out = clone(data || { meta: {}, pages: [] });
    if (!Array.isArray(out.pages)) out.pages = [];
    const applied = [];
    for (const raw of ops || []) {
      if (!raw || typeof raw !== "object") continue;
      const name = String(raw.op || "");
      if (name === "movePage") {
        const from = parseInt(raw.from, 10) - 1;
        const to = parseInt(raw.to, 10) - 1;
        if (from < 0 || from >= out.pages.length || to < 0 || to >= out.pages.length) continue;
        const [p] = out.pages.splice(from, 1);
        out.pages.splice(to, 0, p);
        applied.push(raw);
      } else if (name === "duplicatePage") {
        const idx = parseInt(raw.index, 10) - 1;
        if (idx < 0 || idx >= out.pages.length || out.pages.length >= 12) continue;
        out.pages.splice(idx + 1, 0, markUserEdited(clone(out.pages[idx])));
        applied.push(raw);
      } else if (name === "insertPage") {
        const idx = parseInt(raw.index, 10) - 1;
        const pg = raw.page || blankPage(raw.type);
        const at = Math.max(0, Math.min(idx + 1, out.pages.length));
        out.pages.splice(at, 0, markUserEdited(pg));
        applied.push(raw);
      } else if (name === "deletePage") {
        const idx = parseInt(raw.index, 10) - 1;
        if (idx < 0 || idx >= out.pages.length) continue;
        const t = out.pages[idx].type;
        if (t === "cover" || t === "ending") {
          const count = out.pages.filter((p) => p.type === t).length;
          if (count <= 1) continue;
        }
        out.pages.splice(idx, 1);
        applied.push(raw);
      }
    }
    return { data: out, applied };
  }

  global.XHSOps = {
    clone,
    listKeyForPage,
    applyPageOps,
    applyDocumentOps,
    blankPage,
    normalizeOp,
    markUserEdited,
  };
})(typeof window !== "undefined" ? window : globalThis);

File: backend/ops.py
# -*- coding: utf-8 -*-
"""Structured page/document ops — shared with frontend ops.js."""
from __future__ import annotations

from copy import deepcopy
import re
from typing import Any

from prompts.layout_catalog import apply_recipe_defaults, get_recipe, is_valid_variant
from pipeline import (  # noqa: F401 — reuse fallbacks & helpers
    _PAGE_TYPES,
    _REWRITE_FIELDS,
    _apply_page_density,
    _clone_page,
    _fallback_compact_points,
    _fallback_to_card,
    _fallback_to_free_quote,
    _fallback_to_timeline,
    _list_key_for_page,
    _mark_user_edited,
    _summary_from_ops,
)


def _parse_path(path: str) -> list:
    parts: list = []
    for m in re.finditer(r"([^.\[\]]+)|\[(\d+)\]", str(path or "")):
        if m.group(1):
            parts.append(m.group(1))
        else:
            parts.append(int(m.group(2)))
    return parts


def _get_at_path(obj: Any, path: str) -> Any:
    cur = obj
    for p in _parse_path(path):
        if cur is None:
            return None
        if isinstance(cur, list) and isinstance(p, int):
            if p < 0 or p >= len(cur):
                return None
            cur = cur[p]
        elif isinstance(cur, dict):
            cur = cur.get(p)
        else:
            return None
    return cur


def _set_at_path(obj: Any, path: str, value: Any) -> None:
    parts = _parse_path(path)
    if not parts:
        return
    cur = obj
    for i, p in enumerate(parts[:-1]):
        nxt = parts[i + 1]
        if isinstance(cur, list) and isinstance(p, int):
            while len(cur) <= p:
                cur.append([] if isinstance(nxt, int) else {})
            child = cur[p]
        else:
            child = cur.get(p) if isinstance(cur, dict) else None
        if child is None or not isinstance(child, (dict, list)):
            child = [] if isinstance(nxt, int) else {}
            if isinstance(cur, list) and isinstance(p, int):
                cur[p] = child
            elif isinstance(cur, dict):
                cur[p] = child
        cur = child
    last = parts[-1]
    if isinstance(cur, list) and isinstance(last, int):
        while len(cur) <= last:
            cur.append(None)
        cur[last] = value
    elif isinstance(cur, dict):
        cur[last] = value


def _list_path_spec(op: dict[str, Any]) -> tuple[str, bool] | None:
    lp = str(op.get("listPath") or "").strip()
    if lp:
        return lp, True
    lst = str(op.get("list") or "").strip()
    if lst:
        return lst, "[" in lst
    return None


def _read_list(page: dict, spec: tuple[str, bool]) -> list | None:
    path, nested = spec
    if nested:
        val = _get_at_path(page, path)
        return list(val) if isinstance(val, list) else None
    val = page.get(path)
    return list(val) if isinstance(val, list) else None


def _write_list(page: dict, spec: tuple[str, bool], items: list) -> None:
    path, nested = spec
    if nested:
        _set_at_path(page, path, items)
    else:
        page[path] = items


def _normalize_op(raw: Any) -> dict[str, Any] | None:
    if not isinstance(raw, dict):
        return None
    name = str(raw.get("op") or "").strip()
    allowed = {
        "setDensity",
        "setFontScale",
        "rewriteField",
        "deleteItem",
        "updateItem",
        "insertItem",
        "moveItem",
        "setPageType",
        "setPointsLayout",
        "setPageLayoutVariant",
        "setLayoutRecipe",
        "setImage",
        "setImageFocus",
        "patchPath",
    }
    if name not in allowed:
        return None
    out: dict[str, Any] = {"op": name}
    if name == "setDensity":
        d = str(raw.get("density") or "").lower()
        if d not in ("compact", "normal", "airy"):
            return None
        out["density"] = d
    elif name == "setFontScale":
        try:
            fs = float(raw.get("fontScale"))
        except (TypeError, ValueError):
            return None
        out["fontScale"] = round(max(0.65, min(1.4, fs)), 3)
    elif name == "rewriteField":
        field = str(raw.get("field") or "").strip()
        if field not in _REWRITE_FIELDS:
            return None
        out["field"] = field
        out["value"] = "" if raw.get("value") is None else str(raw.get("value"))
    elif name in ("deleteItem", "updateItem", "insertItem"):
        try:
            idx = int(raw.get("index"))
        except (TypeError, ValueError):
            return None
        if idx < 1:
            return None
        out["index"] = idx
        lst = str(raw.get("list") or "").strip()
        if lst:
            out["list"] = lst
        lp = str(raw.get("listPath") or "").strip()
        if lp:
            out["listPath"] = lp
        if name == "insertItem":
            out["item"] = raw.get("item")
        if name == "updateItem":
            for k in ("head", "body", "value", "text", "time", "label"):
                if k in raw and raw[k] is not None:
                    out[k] = str(raw[k])
            try:
                vi = int(raw.get("valueIndex"))
                if vi >= 1:
                    out["valueIndex"] = vi
            except (TypeError, ValueError):
                pass
    elif name == "moveItem":
        try:
            fr = int(raw.get("from"))
            to = int(raw.get("to"))
        except (TypeError, ValueError):
            return None
        if fr < 1 or to < 1:
            return None
        out["from"] = fr
        out["to"] = to
        lst = str(raw.get("list") or "").strip()
        if lst:
            out["list"] = lst
        lp = str(raw.get("listPath") or "").strip()
        if lp:
            out["listPath"] = lp
    elif name == "patchPath":
        p = str(raw.get("path") or "").strip()
        if not p:
            return None
        out["path"] = p
        out["value"] = raw.get("value")
    elif name == "setImage":
        out["slot"] = str(raw.get("slot") or "image")
        out["url"] = str(raw.get("url") or "")
        if raw.get("focus"):
            out["focus"] = raw["focus"]
    elif name == "setImageFocus":
        out["slot"] = str(raw.get("slot") or "image")
        try:
            out["x"] = max(0, min(100, float(raw.get("x", 50))))
            out["y"] = max(0, min(100, float(raw.get("y", 50))))
        except (TypeError, ValueError):
            out["x"] = 50
            out["y"] = 50
    elif name == "setPageType":
        t = str(raw.get("type") or "").strip()
        if t not in _PAGE_TYPES:
            return None
        out["type"] = t
    elif name == "setPointsLayout":
        layout = str(raw.get("layout") or "").lower()
        if layout not in ("ledger", "cards"):
            return None
        out["layout"] = layout
    elif name == "setPageLayoutVariant":
        variant = str(raw.get("variant") or raw.get("layoutVariant") or "").strip()
        if not variant:
            return None
        out["variant"] = variant
    elif name == "setLayoutRecipe":
        rid = str(raw.get("recipeId") or "").strip()
        if not rid or not get_recipe(rid):
            return None
        out["recipeId"] = rid
    return out


def apply_page_ops(
    page: dict, ops: list[dict[str, Any]]
) -> tuple[dict, list[dict[str, Any]], str]:
    out = _clone_page(page) if isinstance(page, dict) else {"type": "card", "title": "内容", "body": ""}
    applied: list[dict[str, Any]] = []
    for raw in ops or []:
        op = _normalize_op(raw)
        if not op:
            continue
        name = op["op"]
        if name == "setDensity":
            out = _apply_page_density(out, op["density"])
            op = {**op, "fontScale": out.get("fontScale")}
        elif name == "setFontScale":
            fs = op["fontScale"]
            out["fontScale"] = fs
            dens = str(out.get("density") or "normal")
            if fs <= 0.9:
                dens = "compact"
            elif fs >= 1.1:
                dens = "airy"
            out["density"] = dens
            style = out.get("style") if isinstance(out.get("style"), dict) else {}
            out["style"] = {**style, "density": dens, "fontScale": fs}
            op = {**op, "density": dens}
        elif name == "rewriteField":
            out[op["field"]] = op["value"]
        elif name == "patchPath":
            _set_at_path(out, op["path"], op["value"])
        elif name == "deleteItem":
            spec = _list_path_spec(op)
            if not spec:
                key = _list_key_for_page(out)
                if not key:
                    continue
                spec = (key, False)
            items = _read_list(out, spec)
            if items is None:
                continue
            idx = int(op["index"])
            if not (1 <= idx <= len(items)):
                continue
            items = items[: idx - 1] + items[idx:]
            _write_list(out, spec, items)
        elif name == "insertItem":
            spec = _list_path_spec(op)
            if not spec:
                key = _list_key_for_page(out)
                if not key:
                    continue
                spec = (key, False)
            items = _read_list(out, spec) or []
            idx = min(int(op["index"]), len(items) + 1)
            tail = spec[0].split(".")[-1] if spec[1] else spec[0]
            item = op.get("item") or (
                {"head": "新条目", "body": ""} if tail in ("items", "steps", "tips") else "新条目"
            )
            items = items[: idx - 1] + [item] + items[idx - 1 :]
            _write_list(out, spec, items)
        elif name == "moveItem":
            spec = _list_path_spec(op)
            if not spec:
                key = _list_key_for_page(out)
                if not key:
                    continue
                spec = (key, False)
            items = _read_list(out, spec)
            if items is None:
                continue
            fr = int(op["from"]) - 1
            to = int(op["to"]) - 1
            if not (0 <= fr < len(items) and 0 <= to < len(items)):
                continue
            moved = items.pop(fr)
            items.insert(to, moved)
            _write_list(out, spec, items)
        elif name == "updateItem":
            spec = _list_path_spec(op)
            if not spec:
                key = _list_key_for_page(out)
                if not key:
                    continue
                spec = (key, False)
            items = _read_list(out, spec)
            if items is None:
                continue
            idx = int(op["index"])
            if not (1 <= idx <= len(items)):
                continue
            cur = items[idx - 1]
            new_text = op.get("body") or op.get("value") or op.get("text")
            if isinstance(cur, dict):
                nxt = {**cur}
                for k in ("head", "body", "value", "text", "time", "label"):
                    if op.get(k) is not None:
                        nxt[k] = op[k]
                vi_raw = op.get("valueIndex")
                if vi_raw is not None:
                    try:
                        vi = int(vi_raw) - 1
                    except (TypeError, ValueError):
                        vi = -1
                    if vi >= 0:
                        vals = list(nxt.get("values") or [])
                        while len(vals) <= vi:
                            vals.append("")
                        cell_val = op.get("value")
                        if cell_val is None:
                            cell_val = new_text
                        if cell_val is not None:
                            vals[vi] = str(cell_val)
                        nxt["values"] = vals
                if new_text is not None and op.get("body") is None and op.get("value") is None and op.get("text") is None and vi_raw is None:
                    tail = spec[0].split(".")[-1] if spec[1] else spec[0]
                    if "body" in cur or tail in ("items", "steps", "tips"):
                        nxt["body"] = new_text
                    elif "text" in cur:
                        nxt["text"] = new_text
                    else:
                        nxt["body"] = new_text
                items[idx - 1] = nxt
            else:
                items[idx - 1] = new_text if new_text is not None else cur
            _write_list(out, spec, items)
        elif name == "setImage":
            slot = op.get("slot") or "image"
            if isinstance(slot, str) and slot.startswith("images["):
                import re

                m = re.match(r"images\[(\d+)\]", slot)
                if m:
                    i = int(m.group(1))
                    imgs = list(out.get("images") or [])
                    while len(imgs) <= i:
                        imgs.append("")
                    imgs[i] = op["url"]
                    out["images"] = imgs
            else:
                out["image"] = op["url"]
            if op.get("focus"):
                out["imageFocus"] = op["focus"]
        elif name == "setImageFocus":
            out["imageFocus"] = {"x": op["x"], "y": op["y"]}
        elif name == "setPageType":
            new_type = op["type"]
            if new_type != out.get("type"):
                if new_type == "timeline":
                    out = _fallback_to_timeline(out)
                elif new_type == "points":
                    out = _fallback_compact_points(out)
                elif new_type == "quote":
                    out = _fallback_to_free_quote(out)
                    out["type"] = "quote"
                elif new_type == "card":
                    out = _fallback_to_card(out)
                elif new_type == "free":
                    out = _fallback_to_free_quote(out)
                    out["type"] = "free"
                else:
                    out["type"] = new_type
            else:
                out["type"] = new_type
        elif name == "setPointsLayout":
            if out.get("type") != "points":
                continue
            layout = op["layout"]
            out["pointsStyle"] = layout
            style = out.get("style") if isinstance(out.get("style"), dict) else {}
            out["style"] = {**style, "pointsStyle": layout}
            out["layoutVariant"] = layout
            out["style"]["layoutVariant"] = layout
        elif name == "setPageLayoutVariant":
            ptype = str(out.get("type") or "")
            variant = op["variant"]
            if not ptype or not is_valid_variant(ptype, variant):
                continue
            out["layoutVariant"] = variant
            style = out.get("style") if isinstance(out.get("style"), dict) else {}
            out["style"] = {**style, "layoutVariant": variant}
            if ptype == "points" and variant in ("cards", "ledger"):
                out["pointsStyle"] = variant
                out["style"]["pointsStyle"] = variant
        elif name == "setLayoutRecipe":
            out = apply_recipe_defaults(out, op["recipeId"])
        applied.append(op)
    out = _mark_user_edited(out)
    summary = _summary_from_ops(applied)
    return out, applied, summary


def _blank_page(page_type: str | None = None) -> dict[str, Any]:
    t = page_type or "card"
    templates: dict[str, dict[str, Any]] = {
        "cover": {"type": "cover", "title": "新封面", "subtitle": "副标题"},
        "ending": {"type": "ending", "title": "结尾", "desc": "行动号召", "cta": ["收藏备用"]},
        "points": {"type": "points", "title": "要点", "items": [{"head": "1", "body": "内容"}]},
        "timeline": {"type": "timeline", "title": "步骤", "steps": [{"time": "Day1", "head": "步骤", "body": "说明"}]},
        "card": {"type": "card", "title": "卡片", "body": "正文"},
        "quote": {"type": "quote", "text": "金句"},
        "compare": {
            "type": "compare",
            "title": "对比",
            "cols": [{"head": "维度"}, {"head": "A", "tone": "neg"}, {"head": "B", "tone": "pos"}],
            "rows": [{"label": "项", "values": ["", ""]}],
        },
        "summary": {"type": "summary", "title": "总结", "items": ["要点一"]},
        "free": {"type": "free", "title": "展开", "paragraphs": ["正文"]},
        "photo": {"type": "photo", "title": "配图", "body": "说明"},
        "gallery": {"type": "gallery", "title": "图集", "images": []},
    }
    return deepcopy(templates.get(t, templates["card"]))


def apply_document_ops(data: dict, ops: list[dict[str, Any]]) -> tuple[dict, list[dict[str, Any]]]:
    out = deepcopy(data) if isinstance(data, dict) else {"meta": {}, "pages": []}
    if not isinstance(out.get("pages"), list):
        out["pages"] = []
    applied: list[dict[str, Any]] = []
    for raw in ops or []:
        if not isinstance(raw, dict):
            continue
        name = str(raw.get("op") or "")
        if name == "movePage":
            try:
                fr = int(raw.get("from")) - 1
                to = int(raw.get("to")) - 1
            except (TypeError, ValueError):
                continue
            pages = out["pages"]
            if not (0 <= fr < len(pages) and 0 <= to < len(pages)):
                continue
            p = pages.pop(fr)
            pages.insert(to, p)
            applied.append(raw)
        elif name == "duplicatePage":
            try:
                idx = int(raw.get("index")) - 1
            except (TypeError, ValueError):
                continue
            pages = out["pages"]
            if not (0 <= idx < len(pages)) or len(pages) >= 12:
                continue
            pages.insert(idx + 1, _mark_user_edited(deepcopy(pages[idx])))
            applied.append(raw)
        elif name == "insertPage":
            try:
                idx = int(raw.get("index")) - 1
            except (TypeError, ValueError):
                continue
            pg = raw.get("page") or _blank_page(str(raw.get("type") or "card"))
            at = max(0, min(idx + 1, len(out["pages"])))
            out["pages"].insert(at, _mark_user_edited(pg))
            applied.append(raw)
        elif name == "deletePage":
            try:
                idx = int(raw.get("index")) - 1
            except (TypeError, ValueError):
                continue
            pages = out["pages"]
            if not (0 <= idx < len(pages)):
                continue
            t = pages[idx].get("type")
            if t in ("cover", "ending"):
                if sum(1 for p in pages if p.get("type") == t) <= 1:
                    continue
            pages.pop(idx)
            applied.append(raw)
    return out, applied
